### PR TITLE
Add dotenv support

### PR DESCRIPTION
**Pull request by xanimo**
### Motivation
Developing locally is difficult to configure since the app is looking for hardcoded paths for .conf and .ztsd files in /assets directory that are moved to a directory named /app during container build.  

### Solution
By introducing support of .env files we can allow easier manipulation of these paths in addition to other config vars.  Primary motivation for use is making developing locally easier and eventually decoupling the api and dogecoind.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -65,6 +65,7 @@ RUN cd src \
   && cd .. \
   && mv src/rosetta-dogecoin /app/rosetta-dogecoin \
   && mv src/assets/* /app \
+  && mv src/.env.dist /app \
```
Comment (patricklodder) [inline Dockerfile:68]: I think we should rename it `.env`

File: .env.dist [hunk 2]
```diff
@@ -0,0 +1,8 @@
+MODE=ONLINE
+NETWORK=TESTNET
+PORT=8080
+mainnetConfigPath="/app/bitcoin-mainnet.conf"
+testnetConfigPath="/app/bitcoin-testnet.conf"
+testnetTransactionDictionary="/app/testnet-transaction.zstd"
+mainnetTranscationDictionary="/app/mainnet-transaction.zstd"
+DataDirectory="/data"
```
Comment (patricklodder) [inline .env.dist:8]: please add a newline at the end of the file

File: main.go [hunk 3]
```diff
@@ -114,6 +115,10 @@ func startOnlineDependencies(
 }
 
 func main() {
+	if err := godotenv.Load(".env.dist"); err != nil {
```
Comment (patricklodder) [inline main.go:118]: this should be `.env`

File: main.go [hunk 4]
```diff
@@ -114,6 +115,10 @@ func startOnlineDependencies(
...
+		log.Fatal("Error loading .env file", err)
```
Comment (patricklodder) [inline main.go:119]: no need to fatal without the env file, it can be optional

**Review by patricklodder — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 5]
```diff
@@ -65,6 +65,7 @@ RUN cd src \
   && cd .. \
   && mv src/rosetta-dogecoin /app/rosetta-dogecoin \
   && mv src/assets/* /app \
+  && mv src/.env.dist /app/.env \
```
[the remark this answers is not in the document]
Comment (chromatic) [reply Dockerfile:68]: Does Rosetta have any guidelines on how to launch the Docker container? I didn't see any in the linked guide.

If not -- if that's open to implementation guidelines -- then selecting the environment from `run` is feasible.

File: .env.dist [hunk 6]
```diff
@@ -0,0 +1,5 @@
+MODE=ONLINE
```
[the remark this answers is not in the document]
Comment (solo-fish) [reply .env.dist:1]: > I think this could be helpful although I'm not sure it's necessary.

For sure can you help me understand why it's not necessary? We need 1 image that can be run in four different configurations of these values. How will you support that?

File: main.go [hunk 7]
```diff
@@ -130,7 +137,11 @@ func main() {
 
 	logger := loggerRaw.Sugar().Named("main")
 
-	cfg, err := dogecoin.LoadConfiguration(configuration.DataDirectory)
+	if configuration.DataDirectory != "" {
```
[the remark this answers is not in the document]
Comment (solo-fish) [reply main.go:140]: SGTM

File: Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (xanimo) [reply Dockerfile:68]: It's configured to work without a .env file and when you pass -e flags via docker as seen in Makefile e.g. `-e "MODE=OFFLINE" -e "NETWORK=TESTNET" -e "PORT=8081"` they override any existing environment variables in .env.  All in all this is an optional enhancement should anyone want to use it.

File: .env.dist (hunk 6 above)
[the remark this answers is not in the document]
Comment (xanimo) [reply .env.dist:1]: Does my last comment below clear any of this up?

[the remark this answers is not in the document]
Comment (solo-fish) [reply .env.dist:1]: >  All in all this is an optional enhancement should anyone want to use it.

I thought your goal was to start consolidating the different configs we have strewn about our package. 

I am **not aligned** with this _"optional enhancement"_ as you call it; adding more redundant configuration just adds to the problems discussed in #24.

I mean, look at all these current places we have some config values defined;

- [Makefile](https://github.com/rosetta-dogecoin/rosetta-dogecoin/blob/53a1733deba341965fb461705c2c2ba6b0ce6aec/Makefile#L35), 
- [configuration/configuration.go](https://github.com/rosetta-dogecoin/rosetta-dogecoin/blob/main/configuration/configuration.go#L40-L66), 
- [dogecoin/configuration.go](https://github.com/rosetta-dogecoin/rosetta-dogecoin/blob/main/dogecoin/configuration.go#L34-L47), 
- [dogecoin/params.go](https://github.com/rosetta-dogecoin/rosetta-dogecoin/blob/main/dogecoin/params.go#L45-L51), 
- [dogecoin/types.go](https://github.com/rosetta-dogecoin/rosetta-dogecoin/blob/main/dogecoin/types.go#L24-L44), 


I'm of the opinion that all of this configuration should be consolidated & structured in a known location - this PR goes against that and just adds complexity.

I'd ask others to review if you're looking for approval or switch to an approach that aligns with consolidating our configs as per #24.

File: Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (solo-fish) [reply Dockerfile:68]: Yeah you can see my response [here](https://github.com/rosetta-dogecoin/rosetta-dogecoin/pull/40#discussion_r630679523). TLDR: I am not aligned with this approach.

[the remark this answers is not in the document]
Comment (xanimo) [reply Dockerfile:68]: Ok

File: .env.dist (hunk 6 above)
[the remark this answers is not in the document]
Comment (xanimo) [reply .env.dist:1]: Ok.

**Comment by patricklodder** [comment]
@solo-fish 

> I am not aligned with this "optional enhancement" as you call it; adding more redundant configuration just adds to the problems discussed in #24.
> [..]
> - Makefile,
> - configuration/configuration.go,
> - dogecoin/configuration.go,
> - dogecoin/params.go,
> - dogecoin/types.go,

There are 3 types of "configuration":
1. *Dogecoin chain parameters and encoding rules.* Those are rather static for this repository as deviating from dogecoin/dogecoin would essentially create a not-Dogecoin chain. So these can be safely hardcoded. This applies to `params.go`, `genesis.go` and `types.go` and these can stay as-is.
2. *Configuration that changes the behavior of internals.* Many of these are in `configuration.go`, some of them can be overridden with environment variables, some cannot. Defaults can be hardcoded, but ALL should be able to be overridden with environment vars. The question is do we put those in a default `.env` file, or do we put them in the .go source code. We can make a choice here, but I like the .env approach as a "default configuration file" that we can package (all config in there can then be overridden by actual environment vars with the same name) and then build into the standard container. This allows for the tightly coupled, single container, approach that Coinbase wants for rosetta deployments, yet allows other deployments to be much more flexible by creating a .env for whatever the deployment scenario is? 
    - Example: I run ~60 dogecoind deployments right now, some of them fully private for back-end services. Do I really want to run another node inside a massive container? I don't think so. I want to run a container w/ rosetta processes that hook into my existing node fleet. I also want to be able to put a reverse TCP proxy (w/ Kubernetes / haproxy) between my node(s) and be able to failover. Currently I am severely constrained with that because I'd have to change the hardcoded stuff in `bitcoin/*.go`. This PR would be a first step towards enabling much more flexible deployments, while retaining the default build process into one singular container for compliance with Coinbase requirements.
3. *Operational mode selection*: eg testnet/mainnet, online/offline. These are currently mandatory env variables. I don't think they need to be. Having a default operational mode for each, using the same .env file, would be nice. Eg the stuff that is now shorthanded in `Makefile`, but to be honest, I don't necessarily want to use makefiles for deployments.

**Comment by solo-fish** [comment]
> Dogecoin chain parameters and encoding rules. Those are rather static for this repository as deviating from dogecoin/dogecoin would essentially create a not-Dogecoin chain. So these can be safely hardcoded. This applies to params.go, genesis.go and types.go and these can stay as-is.

This SGTM

> Configuration that changes the behavior of internals. Many of these are in configuration.go, some of them can be overridden with environment variables, some cannot. Defaults can be hardcoded, but ALL should be able to be overridden with environment vars. 

IMO coupling environment variables to configuration overrides isn't a great call (emphasis on my opinion..) 

Looking closer at...

> Many of these are in configuration.go, some of them can be overridden with environment variables, some cannot.

How is a new developer supposed to understand this? Are you going to document it? How is somebody supposed to know _what subset of configuration_  values set in `.env` gets overridden by environment variables? What if I want to run all four permutations ([mainnet, online], [mainnet, offline], [testnet online], [testnet, offline]) in a single container where environment variables are shared? How will people clearly understand this configuration hierarchy? 

**Why not just move all configurable values to.. a config?**  -  If you _want to store these in environment variables_ then part of your deployment process can be _writing your environment variables to the configuration file_.

> This allows for the tightly coupled, single container, approach that Coinbase wants for rosetta deployments, yet allows other deployments to be much more flexible by creating a .env for whatever the deployment scenario is?

Sure - I think there are a ton of ways to go about this... I mean all it involves is **_not baking constants into your go binary..._**

I think my opinion is less valuable here though as you folks have more experience, like I mentioned, so if you guys are all aligned that don't hold up on me! Like you stated this will align us with the rosetta-specifications requirements as well.

**Comment by patricklodder** [comment]
Sorry to be slow on this, I get 6 million notifications per day because of Elon-sponsored dogecoin/dogecoin spam, so I have NO idea what I miss.

> what subset of configuration values set in .env gets overridden by environment variables?

ALL. As soon as something is in .env, it out of the box can get overridden with env vars.

> What if I want to run all four permutations ([mainnet, online], [mainnet, offline], [testnet online], [testnet, offline]) in a single container where environment variables are shared? How will people clearly understand this configuration hierarchy?

1. Why would you want to do that?
2. You cannot do this right now?

> How will people clearly understand this configuration hierarchy?

1. Everything that is in .env can be overridden by environment vars.
2. `.env` can be changed
3. Anything that makes sense to be changed and still be Dogecoin MUST be in .env, so we cool?

Is this reasonable?

**Comment by solo-fish** [comment]
@patricklodder @xanimo 
ACK! Approved on my end (Apologies as well on this hold-up!)
> Why would you want to do that?
> You cannot do this right now?

tldr; a _"full deployment"_ is two instances of the same container running within the same "environment" (or really the same subnet). 

One online container to service public-facing traffic, which, is coupled with an offline container that only accepts requests from the subnetwork (ideally only from its "parent" container) - this is how they specifically outlined in the docs I linked above. 

Take a look at these "containers" I have deployed - they share a segregated root filesystem & unprivileged namespace (showing as you asked for an example of multi-container same environment).
```
+-------------------+---------+----------------------+------------+
| mnet-doge-offline | RUNNING | 10.155.6.244 (dog0)  | CONTAINER  |
+-------------------+---------+----------------------+------------+
| mnet-doge-online  | RUNNING | 10.155.6.231 (dog0)  | CONTAINER  |
+-------------------+---------+----------------------+------------+
| tnet-doge-offline | RUNNING | 10.155.6.214 (dog1)  | CONTAINER  |
+-------------------+---------+----------------------+------------+
| tnet-doge-online  | RUNNING | 10.155.6.91 (dog1)   | CONTAINER  |
+-------------------+---------+----------------------+------------+
```
> Is this reasonable?

YES! I'm sorry as well - **I mentioned please don't block on me!** D:

**Review by solo-fish — APPROVED** [review]
ACK!

File: main.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (patricklodder) [reply main.go:140]: @xanimo could you please fix this one?

**Comment by patricklodder** [comment]
> Take a look at these "containers" I have deployed - they share a segregated root filesystem & unprivileged namespace (showing as you asked for an example of multi-container same environment).

That's how I often run things too. When I'd do it with Kubernetes, this would mean I have 4 deployments of the same container image, with different `container.env` specs. On docker compose, I'd add different `environment` content for each, with docker cli I'd start them with differing `-e` arguments.

Summarizing this PR:

1. One use-case MUST need only 1 (generic) image which configuration can be manipulated by passing environment variables to it. This is achievable with this code, at least for starters. ✅ 
2. There SHOULD be a default configuration. That was already there and doesn't change with this, but the .env makes it more configurable. ✅
3. There COULD be a default mode selection. Currently implemented as `TESTNET`, `ONLINE` (we could discuss whether this is better as `MAINNET` instead of `TESTNET`, raise an issue if you think it's better as mainnet, and why.)

This PR makes configuration more user-friendly and gives us a vehicle to pull hard coded configuration out of `.go` files, into environment vars, per 1. Reducing the need for custom code and custom images.

File: dogecoin/configuration.go [hunk 8]
```diff
@@ -69,9 +55,37 @@ const (
...
+	// persistent data.
+	DataDirectory string = os.Getenv("DATA_DIR")
+)
+
 // LoadConfiguration attempts to create a new Configuration
 // using the ENVs in the environment.
 func LoadConfiguration(baseDirectory string) (*configuration.Configuration, error) {
+	if configurationDirectory == "" {
+		configurationDirectory = "/app"
+	}
```
Comment (patricklodder) [inline dogecoin/configuration.go:83]: I think empty string check should be done here consistently with the comment on `main.go`

**Review by patricklodder — CHANGES_REQUESTED** [review]

File: main.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (xanimo) [reply main.go:140]: Yeah I'll be back to my computer shortly.

File: dogecoin/configuration.go (hunk 8 above)
Comment (xanimo) [reply dogecoin/configuration.go:83]: I'm actually running into some errors using len(dir) != 0 and len(dir) == 0. Can you and @solo-fish explain why this is preferred method of checking an empty string? I was reading a thread on this topic and one user found negligible performance benefits of about .01 ns/op. Not to get into syntactic preference per dev or nit as I am understanding but I'm honestly wondering why this is not recommended.

Comment (patricklodder) [reply dogecoin/configuration.go:83]: To me it's just about consistency. If we change it at one place, we can change it at the other.

**Comment by xanimo** [comment]
Sounds good.

File: dogecoin/configuration.go (hunk 8 above)
Comment (xanimo) [reply dogecoin/configuration.go]: For some reason after I switched to checking by length I was failing the configuration_test.go config_path so I added another argument to check whether or not it was testing as seen on line 82 in dogecoin/configuration.go. Not sure how I passed tests before but I've passed all tests and built a container that ran successfully. <3

Comment (solo-fish) [reply dogecoin/configuration.go]: @xanimo Hey! Sorry for the delayed response - @patricklodder is correct on the standard, however,  you should consider using `os` from the stdlib. I could follow up on this as well with a PR.

Excerpt from the documentation
> Package os provides a platform-independent interface to operating system functionality. The design is Unix-like, although the error handling is Go-like; failing calls return values of type error rather than error numbers.

- https://golang.org/pkg/os

**Review by patricklodder — APPROVED** [review]
ACK!
